### ISSUE TITLE
[Proposal][Compile-Time Generics]

### ISSUE THREAD

**Issue by redradist** [dotnet/csharplang#2433]
My proposal is to add possibility to generate **generic** class at **compile-time**, not in run-time.

**_Motivation:_**
1) _Performance Increase_, because _run-time_ should not create a new type
2) _Native compilation_ where generation of new type at _run-time_ is not trivial operation

To make it optionally I suggest to add some **_specific compiler flag_**

```csharp
class Container<T>
{
    ...
    private T t;
    ...
}

class Program
{
    static void Main(string[] args)
    {
        Container<int> cl0;
        Container<float> cl1;
    }
}
```

After compilation example in assembly we would get two generated classes **_Container\<int\>_** and **_Container\<float\>_**

It will simplify deployment for environment where **run-time generic instantiation** is not possible (like Native compilation)

**Comment by YairHalberstadt** [comment]
Such runtimes already take care of generic instantiation for you.

Have you ever tried to use generics in a runtime where they wouldn't work?

**Comment by svick** [comment]
How would it help to move this concern from the runtime to the compiler?

**Comment by redradist** [comment]
@YairHalberstadt
> Such runtimes already take care of generic instantiation for you.
> 
> Have you ever tried to use generics in a runtime where they wouldn't work?

I consider this feature for two main purposes:
1) Performance Increase, because run-time should not create type
2) Native compilation where generation of new type at run-time is not trivial operation

**Comment by redradist** [comment]
@svick 
> How would it help to move this concern from the runtime to the compiler?

I extended description and added motivation for this feature ... take a look

**Comment by YairHalberstadt** [comment]
> Native compilation where generation of new type at run-time is not trivial operation

This is a solved problem.

**Comment by svick** [comment]
@redradist As I understand it, runtimes for platforms where you can't easily generate types at runtime already work by generating distinct types for `Container<int>` and `Container<float>` at AOT compilation time. So I still don't understand how this would help. How is this proposal better than what already exists?

As for performance increase, I don't understand where that would come from either. What exactly does "run-time creating a new type" mean? What makes you think it significantly affects performance?

**Comment by redradist** [comment]
@YairHalberstadt 
> > Native compilation where generation of new type at run-time is not trivial operation
> 
> This is a solved problem.

As workaround for specific platform (like IOS) and by specific tools (like Mono IL transator)

**Comment by YairHalberstadt** [comment]
> And it is not supported by reference run-time as .NET Core

?
.Net Core supports generics just fine

**Comment by redradist** [comment]
@YairHalberstadt 
> > And it is not supported by reference run-time as .NET Core
> 
> ?
> .Net Core supports generics just fine

You lost context of discussion ...
.NET Core does not support compile-time Generics as done by Mono (but Mono makes it not at compile-time of C# code, but at translation phase to specific platform)

**Comment by marksmeltzer** [comment]
I don't think generics are ever done at runtime unless you're using reflection. Even then, they are baked into hard platform specific code by the JIT compiler.

For platforms that don't support JIT, a platform specific precompiler does the work (e.g. IOS).

There definitely shouldn't be anything in the C# language directly correlated with the process of generics compilation...

Regards,

**Comment by YairHalberstadt** [comment]
CoreCLR and .Net framework have probably the most performant implementation of generics of any platform (compared to Java, c++, etc.). This is because they are able to reuse code between instantiations, whilst at the same time avoiding boxing, and retaining type safety and runtime type information.

Everything about generics in all the supported runtimes *just works* without you having to know anything about it.

There is literally nothing here that needs fixing.

**Comment by scalablecory** [comment]
@YairHalberstadt small code size does not necessarily equal performance.

**Comment by HaloFour** [comment]
How would you suggest this form of generics work across assembly boundaries?

**Comment by msedi** [comment]
Compile time generics would solve the problem of math operations on generics. But I think there is more on this. I could somehow imagine that you need define which generics should be created during runtime. So for example, I could define some sort of a user defined array class

```
  public abstract class MyArray<T> {
    private readonly T[] Data;

    public ref this [int index] => ref Data[index];

    public MyArray (int size) {
      Data = new T[size];
    }

    public static implicit MyArray<T> operator + (MyArray<T> a, MyArray b) {
      if (a.Data.Length != b.Data.Length)
        throw new Exception ();

      MyArray<T> c = new MyArray (a.Data.Length);
      for (int i = 0; i < a.Data.Len; i++)
        c.Data[i] = a.Data[i] + b.Data[i];

      return c;
    }
  }
```

Currently this is not possible, because the different generic types and there is obviously no chance that the jitter is able to recognize that an operation exists (as it is well known).
But if you could help the compiler so that the compiler could generate one variant during compilation the compiler could solve the problem and throw an error if no operation exists, or some other problems occur. For example doing something like this:

`using MyArray<int>;`

could hint the compiler to create the type. Of course the type should be exposed as a variant of the generic type, so all the reflection methods should also work with this behavior. I don't know how the runtime manages the different generic variants internally. This would be much better than duplicating the code manually as we have to do it now.

Additionally, since this is mostly not enough for implementations that require a different method behavior internally. FFT with MKL is an example. If you want to write an FFT wrapper you can have two different types: float and double. I'm thinking of something like:

```
  public abstract class FFT<T> where T float,double
    {
        public abstract T[] ForwardTransform(T[] data);
    }

    public  class FFT<float> : FFT<float>
    {
        public override float[] ForwardTransform(float[] data)
        {
            FFTConfig cfg = new FFTConfig();
            cfg.Precision = PrecisionType.Float;
            ...
        }
    }

    public  class FFT<double> : FFT<double>
    {
        public override double[] ForwardTransform(double[] data)
        {
            FFTConfig cfg = new FFTConfig();
            cfg.Precision = PrecisionType.Float;
            ...
        }
    }
```

Here both required types are already created, there's no need to create additional variants, since it's not even possible with the MKL. There is also no need to write something like this:

```
using MyArray<float>;
using MyArray<double>;
```

One more thing does not exist as you can see with the abstract class, you cannot limit to special types.
Nevertheless, with the above example some types are created during compile time and the runtime only needs to deal with the case that some types of the generic variant have been already created. 

You might say, that I can currently do the same by writing

```
  public abstract class FFT<T> 
    {
        protected readonly Type DataType = typeof(T);

        public abstract T[] ForwardTransform(T[] data);
    }

    public  class FFT_Float : FFT<float>
    {
        public override float[] ForwardTransform(float[] data)
        {
            if (typeof(T) != DataType)
                throw new Exception();

            FFTConfig cfg = new FFTConfig();
            cfg.Precision = PrecisionType.Float;
            ...
        }
    }

    public  class FFT_Double : FFT<double>
    {
        public override double[] ForwardTransform(double[] data)
        {
            if (typeof(T) != DataType)
                throw new Exception();

            FFTConfig cfg = new FFTConfig();
            cfg.Precision = PrecisionType.Float;
            ...
        }
    }
```
But I think this is not the right approach. In general I would say for those types that require a specific behavior and can be created at compile time should be possible, if the language or the runtime does not support the examples above or others.

The question from @HaloFour was, 
> how do you expose across assembly boundaries

I think you don't have to if the runtime can deal with the already created types, they need to seamlessly integrate into all the other generic variants of the class and should not be treated differently. 

If the compile time generics type variants integrate into the type system/reflection system also creating those types at runtime would work. For example the MyArray class above could be created like this.

```
Activator.CreateInstance(typeof(MyArray<>).MakeGenericType(typeof(float)), 10);
Activator.CreateInstance(typeof(MyArray<>).MakeGenericType(typeof(int)), 10);
```
should then work. The first one should fail though, because the runtime cannot handle it, but the second should succeed, because it is already known and a compiled type.

I have currently no clue how the type system and the runtime can handle this, that's the reason why we need you ;-)

**Comment by HaloFour** [comment]
@msedi 

> Currently this is not possible, because of the different generic types and there is obviously no chance that the jitter is able to recognize that an operation exists

The JIT isn't why operators aren't allowed on generic types.  The example you posted is strictly a language limitation.  To the runtime a C# operator is just a static method, and there's no limitation on static methods being generic.

The only limitation that does apply from the runtime is that generic constraints can only be interfaces or base classes.  Invocation of the instance members can be done through a normal virtual call, but there's no facility to do that for static calls.  `T.op_Addition` doesn't resolve to anything at runtime.  There are numerous proposals being tossed around that seek to solve this issue by allowing static members to be delegated through instance members on witness structs.

> If you want to write an FFT wrapper you can have two different types: float and double. I'm thinking of something like

How does that work if `FFT` is `sealed`?  The compiler couldn't roll its own any more than you could, and those types would be incompatible with the CLR-generated types.The CLR already does that form of specialization with structs, so I don't see what the point would be of having the compiler do it.

> I think you don't have to if the runtime can deal with the already created types

My question is how do you define these generic types in other assemblies, and then specialize them in your own?  You couldn't do `using MyArray<double>` in a separate assembly.

> For example the MyArray class above could be created like this.

That already has a defined behavior today.

> The first one should fail though, because the runtime cannot handle it, but the second should succeed, because it is already known and a compiled type.

And that ain't it.  The runtime would know nothing about these compiler-generated derived specializations.  `MyArray<double>` is not compatible with `MySpecializedArray : MyArray<double>`.

> I have currently no clue how the type system and the runtime can handle this, that's the reason why we need you ;-)

In short, it can't.  Runtime changes are exceptionally expensive, so any benefits would have to be similarly exceptional to justify it.

**Comment by theunrepentantgeek** [comment]
@msedi It's already possible for generic classes to work performantly with numerics - it just requires a lot of work on the part of the author of the generic type.

For example, check out the [`Vector<T>`](https://github.com/dotnet/corefx/blob/4d6cde1be300d1fb8b3aec2356d18f1f3d071913/src/Common/src/CoreLib/System/Numerics/Vector.cs) type. It works optimistically with a small set of types because support is hard coded. Have a read - the JIT compiler recognizes the pattern and generates optimal code.

**Comment by msedi** [comment]
> The example you posted is strictly a language limitation

Understood, but what makes it so hard to allow this in the language?

> The only limitation that does apply from the runtime is that generic constraints

If I look closely at the generated IL code for static operators and operators for primitive types, I think there is a difference. For the primitive types, you can simply call ADD whereas for the static operator you call op_Addition. How can this be solved since then there would we two variants of IL code for the same purpose, right?

> How does that work if FFT is sealed

How does the runtime currently take care of these types. In primitive words I was assuming that internally the runtime always creates different code variants for different generic types, so to say like a macro as C++ does it, but more dynamic. My code example is for sure not complete and needs more discussion.

> You couldn't do using MyArray<double> in a separate assembly

The `using MyArray<double>`  was only a quick placeholder to make a type visible and created for the runtime, I was aware that this is a misuse of an already existing behavior.
But why can't you do something like `using MyArray<double>`  in a different assembly. I'm assuming (I might be wrong) that there is a list in the runtime internally that keeps track of all created types. Let's make the example for `MyArray<T>`. Any time I write in the code `var x = new MyArray<int>(10)` I hope the runtime has stored the jitted code for `MyArray<int>(10)` and only needs to create a new instance of MyArray, right? So for all types I have ever wrote `MyArray<int>(10)` or `MyArray<float>(10)` one jitted code variant for each generic type (here: int, float) exists, right? So when goind to other assemblies the runtime also recognizes that these type already exists, so I don't quite understand what's the problem here.

> And that ain't it.

For the FFT you may be right that the syntax is maybe not the one that should be used. In C++ this is template specialisation, so there might be a better way to write it so that the runtime recognizes is as a type of FFT<T> and not as a derivative of FFT<T>.

To summarize, I really don't need or want a runtime change if it could be done with current methods, but I think all agree that I can currently not do operations on generics as it can be done with C++. In fact I can do it, but with efforts that are not negliable and are also error prone since it means that I have to duplicate a lot of code which can be seen in Math and MathF.

Just to show a primitive example:

   ```
     /// <summary>
        /// C = A + B
        /// </summary>
        /// <param name="A"></param>
        /// <param name="B"></param>
        /// <param name="C"></param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static T[] Add<T>(T[] A, T[] B, T[] C = null, ArrayRestriction Restriction = null) where T : struct
        {
            int LastIndex = ArrayHelperInternal.Init(A, B, ref C, ref Restriction);

            for (int i = Restriction.Offset; i < LastIndex; i++)
                C[i] = MathIL.Add(A[i], B[i]);

            return C;
        }
```

The method only works, because I has to work around C# and implement the add in IL code. the additional limitation is that I'm not able to limit the type to numeric data types (as also requested a lot).
One suggestion from @theunrepentantgeek was to do it like this:

```
 		/// <summary>
        /// C = A + B
        /// </summary>
        /// <param name="A"></param>
        /// <param name="B"></param>
        /// <param name="C"></param>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static T[] Add<T>(T[] A, T[] B, T[] C = null, ArrayRestriction Restriction = null) where T : struct
        {
            int LastIndex = ArrayHelperInternal.Init(A, B, ref C, ref Restriction);

            switch(Type.GetTypeCode(typeof(T))
            {
                case TypeCode.Byte:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((byte)(object)A[i] +  (byte)(object)B[i]);
                	break;
                case TypeCode.UInt16:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((ushort)(object)A[i] +  (ushort)(object)B[i]);
                	break;
                case TypeCode.Int16:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((short)(object)A[i] +  (short)(object)B[i]);
                	break;
                case TypeCode.UInt32:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((uint)(object)A[i] +  (uint)(object)B[i]);
                	break;
                case TypeCode.Int32:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((int)(object)A[i] +  (int)(object)B[i]);
                	break;
                case TypeCode.UInt64:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((ulong)(object)A[i] +  (ulong)(object)B[i]);
                	break;
                case TypeCode.Int64:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((long)(object)A[i] +  (long)(object)B[i]);
                	break;
                case TypeCode.Single:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((float)(object)A[i] +  (float)(object)B[i]);
                	break;
                case TypeCode.Double:
            		for (int i = Restriction.Offset; i < LastIndex; i++)                        
                		C[i] = (T)(object)((double)(object)A[i] +  (long)(double)B[i]);
                	break;
            }
                   
            return C;
        }
```

I'm really not sure if this is the right approach. Especially when it comes to the boxing that is happening here for both arguments. Compared to my approach with IL (which I don't like) and the "extended" example there is 54 lines of code vs. 16 lines (which is of course not the measure), but my method needs to be done correctly only once, whereas in example 2 I have to do a lot more and for each type I can introduce bugs. Keep in mind that there is not only Add, there is Abs, Log, Sub, Min, etc. so for all these operations I have to repeat the code on and on a again.

**Comment by HaloFour** [comment]
@msedi 

> Understood, but what makes it so hard to allow this in the language?

They're already looking into allowing stuff like this into the language:

https://github.com/dotnet/csharplang/issues/1711
https://github.com/dotnet/csharplang/issues/164
https://github.com/dotnet/csharplang/issues/110

The generic math examples are on their list.

> If I look closely at the generated IL code for static operators and operators for primitive types, I think there is a difference. For the primitive types, you can simply call ADD whereas for the static operator you call op_Addition. How can this be solved since then there would we two variants of IL code for the same purpose, right?

You rely on the runtime to specialize for structs and inline the operator methods.  It's pretty good at it.

> How does the runtime currently take care of these types.

The runtime can emit any number of native versions for a generic type.  It's not held to the same language and metadata rules as dictated by IL.  There's no way for C# to emit a specialized version without actually emitting a derived class, overriding `virtual` methods.  And in doing that there's no way for the runtime to accept that the derived class is the generic class specialized for a specific generic type argument.  Nor is there a way to "specialize" anything non-virtual or private.  You'd need runtime changes for all of the above. 

> But why can't you do something like `using MyArray<double>` in a different assembly.

That's what I was asking.  If you define `MyArray<T>` in one assembly and reference it in another you lose all of the source that went into defining `MyArray<T>`.  What is there for C# to specialize in your other assembly?  It can't generate a specialized flavor of the source, it doesn't have it.  At best it might have IL, but that's not necessarily true either.  What if that other assembly was written in VB.NET?

> So when goind to other assemblies the runtime also recognizes that these type already exists, so I don't quite understand what's the problem here.

The runtime does nothing like that today.  It keeps track of its own specialization of a generic type, assuming it makes any at all.  That the runtime might make specializations at all is an implementation detail and not a part of the spec for the CLR.  So that's more runtime changes.

**Comment by johnkellyoxford** [comment]
> Motivation:
> 
> Performance Increase, because run-time should not create a new type

No. It is a one time, extremely well optimised operation that minimises code size on disk while allowing proper generic behaviour. Every instance of a generic that is a reference type shares the same code. If you have extensively profiled your program and found for some insane reason that the instantiation of certain generics are causing a major performance bottleneck by some magical means, just force them to be eagerly instantiated on app loadup. Until you show an example of where this is causing you performance issues regularly, it just doesn't make sense. You could always handwrite out the specialised types for hot paths if necessary.

> Native compilation where generation of new type at run-time is not trivial operation
> To make it optionally I suggest to add some specific compiler flag

Refer to above

> After compilation example in assembly we would get two generated classes Container<int> and Container<float>

So we get to massively bloat the size of the binary for no real benefits? Great

> It will simplify deployment for environment where run-time generic instantiation is not possible (like Native compilation)

No it won't. Native compilation can just do this with normal generics. I think you don't understand how native compilation of IL works. This would not be a C# compile time thing, it would be an IL compiler thing. Which is not a language change

**Comment by gafter** [comment]
This proposal does not appear to suggest any change to the C# programming language.  It is merely suggesting an implementation technique.  As the author points out, at least one platform uses this technique and at least one does not.  So there is no affect on the language.

As such, this would be an implementation suggestion for some implementation.  Which implementation are you suggesting should adopt this technique?  Why do you believe that file I/O to read larger assemblies is likely to be faster at runtime than the JIT producing these types in memory as needed?  Is it just a guess or do you have some hard evidence to support the suggestion?  What IL changes do you suggest to support the proposed change (I don't think it is expressible in IL today)?

**Comment by redradist** [comment]
> 
> 
> This proposal does not appear to suggest any change to the C# programming language. It is merely suggesting an implementation technique. As the author points out, at least one platform uses this technique and at least one does not. So there is no affect on the language.
> 
> As such, this would be an implementation suggestion for some implementation. Which implementation are you suggesting should adopt this technique? Why do you believe that file I/O to read larger assemblies is likely to be faster at runtime than the JIT producing these types in memory as needed? Is it just a guess or do you have some hard evidence to support the suggestion? What IL changes do you suggest to support the proposed change (I don't think it is expressible in IL today)?

Perfomance is just my guess, but it would be good for generalized solution for all platforms
For example, this technics would be great for IoT device, because to make JIT on IoT devices is very tricky
For example we have this technic for IoT and I have created IoT library in .dll (not in IoT binary format) and I want to share this library for all platform without recompiling, how can I do this ?

**Comment by gafter** [comment]
The C# specification rarely specifies performance or anything about platforms, or how a .dll or any artifact is produced from a C# program.  That is entirely out of scope for the language specification.

**Comment by gafter** [comment]
@marksmeltzer 
> I don't think generics are ever done at runtime unless you're using reflection.

Yes, they are done at runtime.  Otherwise the following program could not run.

``` c#
public class Program
{
    public static void Main()
    {
        R<int>(60);
    }

    private static void R<T>(int i)
    {
        if (i <= 0) return;
        R<G<T>>(i - 1);
    }
}

class G<T> { }
```

**Comment by marksmeltzer** [comment]
@gafter wouldn't the JITTER prebake closed types?

I get how that scenario you depicted is open - and the runtime definitely can handle that since the ad hoc reflection scenarios work. For closed types, I am under the impression that the JITTER already compiles them in a manner consistent with the original request. 

I actually have some catching code for data pages that is doing something similar to the scenario you presented 😂
